packages/playground/blueprints/src/lib/steps/apply-wordpress-patches/wp-content/mu-plugins/0-playground.php

```diff
@@ -0,0 +1,112 @@
+<?php
+/**
+ * Add a notice to wp-login.php offering the username and password.
+ */
+add_action(
+	'login_message',
+	function () {
+		return <<<EOT
+<div class="message info">
+	<strong>username:</strong> <code>admin</code><br /><strong>password</strong>: <code>password</code>
+</div>
+EOT;
+	}
+);
+
+/**
+ * Because the in-browser Playground doesn't have access to the internet, 
+ * network-dependent features like directories don't work. Normally, you'll 
+ * see a confusing message like "An unexpected error occurred." This mu-plugin
+ * makes it more clear that the feature is not yet supported.
+ * 
+ * https://github.com/WordPress/wordpress-playground/issues/498
+ */
+add_filter('plugins_api_result', function ($res) {
+	if ($res instanceof WP_Error) {
+		$res = new WP_Error(
+			'plugins_api_failed',
+			'Enable networking support in Playground settings to access the Plugins directory. Network access is an <a href="https://github.com/WordPress/wordpress-playground/issues/85">experimental, opt-in feature</a>. If you don\'t want to use it, you can still upload plugins or install them using the <a href="https://wordpress.github.io/wordpress-playground/query-api">Query API</a> (e.g. ?plugin=coblocks).'
+		);
+	}
+	return $res;
+});
+
+add_filter('gettext', function ($translation) {
+	// There is no better hook for swapping the error message
+	// on the themes page, unfortunately.
+	global $pagenow;
+
+	// Only change the message on /wp-admin/theme-install.php
+	if ('theme-install.php' !== $pagenow) {
+		return $translation;
+	}
+
+	if ($translation === 'An unexpected error occurred. Something may be wrong with WordPress.org or this server&#8217;s configuration. If you continue to have problems, please try the <a href="%s">support forums</a>.') {
+		return 'Enable networking support in Playground settings to access the Themes directory. Network access is an <a href="https://github.com/WordPress/wordpress-playground/issues/85">experimental, opt-in feature</a>. If you don\'t want to use it, you can still upload themes or install them using the <a href="https://wordpress.github.io/wordpress-playground/query-api">Query API</a> (e.g. ?theme=pendant).';
+	}
+	return $translation;
+});
+
+
+
+/**
+ * Links with target="top" don't work in the playground iframe because of
+ * the sandbox attribute. What they really should be targeting is the
+ * playground iframe itself (name="playground"). This mu-plugin rewrites
+ * all target="_top" links to target="playground" instead.
+ * 
+ * https://github.com/WordPress/wordpress-playground/issues/266
+ */
+add_action('admin_print_scripts', function () {
+	?>
+	<script>
+		document.addEventListener('click', function (event) {
+			if (event.target.tagName === 'A' && ['_parent', '_top'].includes(event.target.target)) {
+				event.target.target = 'wordpress-playground';
+			}
+		});
+	</script>
+	<?php
+});
```

Comment:
is there enough of a warrant to move this into the service worker? or would that leave out CLI Playgrounds?

or perform it on render output from PHP to ensure that we don't miss it?

Reply:
This is only relevant for the web version so you're on to something here.

The main challenge with doing it outside of WordPress, like in the service worker or the PHP rendering pipeline, is deciding whether the output is meant to be a document displayed by the browser. There may be static HTML files, XHR queries, HTML-like files, CSS files, binary blobs etc, but we only ever want to add this code snippet to UI rendered by WordPress.